File: methods/secant.py
import numpy as np
from math import *

def f(x):
    return x**x - 100;

def secant(f, p_0, p_1, tol, n):
    e_abs = abs(p_1 - p_0);
    
    print('iteration:',0,'p = ', p_0);
    print('iteration:',1,'p = ', p_1, 'error = ', e_abs);
    
    i = 2;
    while i <= n:
        if f(p_1) == f(p_0): #division by cero
            print('solution not found (error in the initial values)');
            break;
        
        p_2 = p_0 - (f(p_0)*(p_1 - p_0))/(f(p_1) - f(p_0)); # secant method formula
        e_abs = abs(p_2- p_1);
        print('iteration:',i,'p = ', p_2, 'error = ', e_abs);
        
        if e_abs < tol: #stop criterion
            print('solution found in x = ', p_2, 'iterations: ', i);
            break;
        
        p_0 = p_1;
        p_1 = p_2;
        
        i+=1;
    if i > n:
        print("solution not found for tolerance:" , tol,"spend iterations:", i-1);
        
secant(f,3,3.2,10**-4,50);

File: methods/matlab-methods/secant.m
%This program finds the solution to the equation f(x)=0 using the method of the secant

%Inputs:
%f, continuous function
%x0, initial approximation
%x1, initial approximation
%tol, tolerance
%Nmax, maximum number of iterations

%Output:
%x, solution
%iter, number of iterations
%err, error

function [x,iter,err]=C6_secante(f,x0,x1,tol,Nmax)

    %initialization
    f0=f(x0);
    f1=f(x1);
    E=1000; 
    cont=1;
    
    %Loop
    while E>tol && cont<Nmax
      xact=x1-f1*(x1-x0)/(f1-f0);
      fact=f(xact);
      E=abs(xact-x1);
      cont=cont+1; 
      x0=x1;
      f0=f1;
      x1=xact;
      f1=fact;
    end
    
    %Delivery of results
    x=xact;
    iter=cont;
    err=E;
    end 